tests/PySys/software-management-end-to-end/environment_sm_management.py

```diff
@@ -0,0 +1,180 @@
+
+        auth = bytes(f"{tenant}/{user}:{password}", "utf-8")
+        self.header = {
+            b"Authorization": b"Basic " + base64.b64encode(auth),
+            b"content-type": b"application/json",
+            b"Accept": b"application/json",
+        }
+
+    def trigger_action(self, package_name, package_id, version, url, action):
+        """Trigger a installation or deinstallation of a package"""
+
+        self.trigger_action_json(
+            [
+                {
+                    "id": package_id,
+                    "name": package_name,
+                    "version": version,
+                    "url": url,
+                    "action": action,
+                }
+            ]
+        )
+
+    def trigger_action_json(self, json):
+        """Take an actions description that is then forwarded to c8y"""
+
+        url = "https://thin-edge-io.eu-latest.cumulocity.com/devicecontrol/operations"
+
+        payload = {
+            "deviceId": self.project.deviceid,
+            "description": "Apply software changes, triggered from PySys test",
+            "c8y_SoftwareUpdate": json,
+        }
+
+        req = requests.post(url, json=payload, headers=self.header)
+
+        self.log.info(f"Response: {req}")
+        self.log.info(f"Response to action: {req.text}")
+
+        if req.status_code != 201:  # Request was accepted
+            raise SystemError("Got HTTP status %s", req.status_code)
+
+    def is_status_fail(self):
+        return self.is_status("FAILED")
+
+    def is_status_success(self):
+        return self.is_status("SUCCESSFUL")
+
+    def is_status(self, status):
+        """Check if the last operation is successfull"""
+
+        timeslot = 600
+        time_to = datetime.now(timezone.utc).replace(microsecond=0)
+        time_from = time_to - timedelta(seconds=timeslot)
+
+        assert is_timezone_aware(time_from)
+
+        date_from = time_from.isoformat(sep="T")
+        date_to = time_to.isoformat(sep="T")
+
+        params = {
+            "deviceId": self.project.deviceid,
+            "pageSize": self.PAGE_SIZE,
+            "dateFrom": date_from,
+            "dateTo": date_to,
+            "revert": "true",
+        }
+
+        url = "https://thin-edge-io.eu-latest.cumulocity.com/devicecontrol/operations"
+        req = requests.get(url, params=params, headers=self.header)
+
+        if req.status_code != 200:  # Request was accepted
+            raise SystemError("Got HTTP status %s", req.status_code)
+
+        jresponse = json.loads(req.text)
+
+        if not jresponse["operations"]:
+            self.log.error("No operations found")
+            return None
+        operation = jresponse["operations"][0]
+
+        self.log.info(operation["status"])
+        # Observed states: PENDING, SUCCESSFUL, EXECUTING, FAILED
+
+        return operation["status"] == status
+
+    def wait_until_succcess(self):
+        """Wait until c8y reports a success
+        TODO This might block forever
```

Comment:
I agree with this concern. We can add timeout and give it to an argument. The timeout depends on package size, the number of packages to apply the change...